src/storage/engine/mod.rs

```diff
@@ -101,6 +103,24 @@ pub trait Engine: Send + Debug {
         }
     }
 
+    fn future_snapshot(
+        &self,
+        ctx: &Context,
+    ) -> Box<Future<Item = Box<Snapshot>, Error = Error> + Send> {
+        let (callback, future) = util::future::gen_paired_future_callback();
+        let val = self.async_snapshot(ctx, callback);
+        if let Err(e) = val {
+            box future::err(e)
+        } else {
+            box future
+                // map future::oneshot::Canceled to Error::Other
+                .map_err(Error::from)
```

Comment:
How about ` Error::Other(box_err!(e))` ?

Reply:
Such errors are handled in the later `flatten`. The `.map_err(Error::from)` here only handles `future::oneshot::Canceled` because the `oneshot` interface returns a `Future<Result<...>, Canceled>`.

Reply:
I think the `flatten` is a little bit misused, try

```rust
box future
    .map_err(|e| Error::Other(box_err!(e)))
    .then(|resp| match resp {
        Ok((_, Ok(snapshot))) => Ok(snapshot),
        Ok((_, Err(e))) | Err(e) => Err(e),
    })
```

Reply:
We can just `.and_then(|x| x)` (which is equivalent to `flatten`).